cisco_ise/actions/lib/base.py

```diff
@@ -0,0 +1,458 @@
+        self.base_api_resource = kwargs.pop('base_api_resource', None)
+        # If the base_api_resource already has a preceeding '/', it should be removed.
+        if self.base_api_resource.startswith('/'):
+            self.base_api_resource = self.base_api_resource[1:]
+        # Check that the HTTP method we are attempting to use is in the list of valid methods
+        if not self.base_api_resource:
+            raise ValueError(
+                'base_api_resource is a required parameter to build the URI for the API Request. '
+                f'base_api_resource recieved: {self.base_api_resource}')
+
+        # if present, the ID for the API resource
+        self.resource_id = kwargs.pop('resource_id', None)
+        # Does not require validation because not all actions will have this
+
+        # if present, the Payload for the request (POST/PUT)
+        self.payload = kwargs.pop('payload', None)
+        # Does not require validation because not all actions will have this, and
+        # JSON Schema is used on the action
+
+        # https://developer.cisco.com/docs/identity-services-engine/#!searching-a-resource
+        # extract the filter operator (AND OR)
+        self.filter_type = kwargs.pop('filter_type', None)
+        self.filter_list = kwargs.pop('filter_list', None)
+
+        # https://developer.cisco.com/docs/identity-services-engine/#!searching-a-resource/sorting
+        self.sort_order = kwargs.pop('sort_order', None)
+        self.sort_field = kwargs.pop('sort_field', None)
+
+        # ISE Supports Payloads and responses in either JSON or XML,
+        # if not specified set both to json.
+        # https://developer.cisco.com/docs/identity-services-engine/#!request-headers/request-headers
+        self.response_format = kwargs.pop('response_format', 'json')
+        self.payload_format = kwargs.pop('payload_format', 'json')
+
+        # Assign kwargs to a class object so its easy to manipulate/reference later
+        self.action_kwargs = kwargs
+
+        return True
+
+    def _prepare_uri(self):
+        """ Parses data into a URI String that can be used by the API client
+        """
+
+        try:
+            self.connection
+        except NameError:
+            raise NameError(
+                '_setup_connection() is required to be ran prior to _prepare_uri()')
+
+        # Check if protocol should be HTTP, if not set it to HTTPS
+        if self.connection['use_http']:
+            protocol = 'http'
+        else:
+            protocol = 'https'
+
+        # if we have a resource_id, we need to add it into the URI after the base_api_resource
+        if self.resource_id:
+            api_resource = f'{self.base_api_resource}/{self.resource_id}'
+        else:
+            api_resource = f'{self.base_api_resource}'
+
+        # construct the base uri
+        _h = self.connection['hostname']
+        self.request_uri = f'{protocol}://{_h}:{HTTPS_PORT}/ers/config/{api_resource}'
```

Comment:
If protocol is `http`, does this need a different port?

Reply:
good catch. Fixing.